proposals/20240606-jwt-handling.md

```diff
@@ -0,0 +1,125 @@
+As of Kubeflow 1.8 the user information has been injected into requests as the `kubeflow-userid` header, from the AuthService (replaced by `oauth2-proxy`). For this approach to be secure there are the following patterns that Kubeflow follows:
+1. Backends in the `kubeflow` namespace that need to know the user identity will rely on `kubeflow-userid` headers in http requests
+2. The AuthService will be adding the `kubeflow-userid` header to all authenticated requests
+3. Only requests from the Istio IngressGateway should be trusted to have the `kubeflow-userid` header
+    1. Backends that shouldn't be exposed to user namespaces (i.e. jupyter-web-app) are only reachable via the Istio IngressGateway
+    2. The KFP backend [explicitly drops requests](https://github.com/kubeflow/manifests/blob/96ce068e16b2a707464471bddc0d2a58e403d1fc/apps/pipeline/upstream/base/installs/multi-user/istio-authorization-config.yaml#L37) from user namespaces if they have this header
+    3. For in-cluster Pods that want to talk to `kubeflow` workloads, which understands identity, the Pods will need to [use a K8s ServiceAccount Token](https://www.kubeflow.org/docs/components/pipelines/v1/sdk/connect-api/#full-kubeflow-subfrom-inside-clustersub)
+### Limitations
+- Not able to express `AuthorizationPolicies` for group header in Istio
+
+The above limitation has pushed us to inject `id_tokens` to requests, via `oauth2-proxy`, to all authenticated requests. This will enable us to define policies in the future for better handling of groups.
+https://istio.io/latest/docs/tasks/security/authorization/authz-jwt/
+
+But the above creates the following issues:
+1. Ambiguity on how to handle different `id_tokens` (i.e. from Dex and from K8s)
+2. Information of user is both in `kubeflow-userid` and in `id_token` of http request
+3. It's not clear if backends should be validating the JWT tokens (i.e. KFP right now validates ServiceAccount tokens [`1`](https://github.com/kubeflow/pipelines/blob/2.2.0/backend/src/apiserver/auth/authenticator_token_review.go#L47-L58) [`2`](https://github.com/kubeflow/pipelines/blob/2.2.0/backend/src/apiserver/resource/resource_manager.go#L1698-L1699) )
+## Spec
+This proposal aims to provide a uniform way for all backends to handle identity tokens, and agreeing on which levels of the stack should responsible for different parts.
+
+This spec proposes to standardise on the following high level agreement, for new backends:
+1. Requests hitting `kubeflow` apps, which expect user identity in requests, should have a JWT in `Authorization: Bearer <token>` header
+2. The service-mesh is responsible for validating the JWTs
+    1. The service-mesh must drop (401) a request if the JWT is invalid (`RequestAuthentication`)
+    2. The service-mesh must drop (403) a request if the JWT is not present, and the application expects requests to have a user identity (`AuthorizationPolicy`)
+3. The backends are not responsible for validating the JWTs or their existence
```

Comment:
The above indeed sound like interesting use cases for the future, but would also like us to evaluate this with a concrete proposal once the need will be more prominent for the above.

The argument for abstracting the service mesh especially is quite interesting, although would require us to iron out a lot of details.

Are you OK with keeping this comment open then, so that we reference it in the future discussions, but not block this PR on the above for now?

Reply:
Sounds good man, let's do it like that.